File: app/src/main/java/com/vtrack/feature/maintenance/types/MaintenanceTypesScreen.kt
package com.vtrack.feature.maintenance.types

import androidx.compose.animation.animateColorAsState
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Build
import androidx.compose.material.icons.filled.Delete
import androidx.compose.material.icons.filled.Edit
import androidx.compose.material3.Badge
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.FloatingActionButton
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.SnackbarDuration
import androidx.compose.material3.SnackbarHost
import androidx.compose.material3.SnackbarHostState
import androidx.compose.material3.SnackbarResult
import androidx.compose.material3.SwipeToDismissBox
import androidx.compose.material3.SwipeToDismissBoxValue
import androidx.compose.material3.Text
import androidx.compose.material3.rememberSwipeToDismissBoxState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.navigation.NavController
import com.vtrack.util.FormatUtil
import com.vtrack.util.MaintenanceStatus
import com.vtrack.util.MaintenanceUrgency
import kotlinx.coroutines.launch
import kotlin.math.abs

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun MaintenanceTypesScreen(
    navController: NavController,
    onAddType: (Long) -> Unit,
    onEditType: (Long, Long) -> Unit,
    onViewHistory: (Long) -> Unit,
    viewModel: MaintenanceTypesViewModel = hiltViewModel()
) {
    val uiState by viewModel.uiState.collectAsState()
    val snackbarHostState = remember { SnackbarHostState() }
    val scope = rememberCoroutineScope()

    Scaffold(
        snackbarHost = { SnackbarHost(snackbarHostState) },
        floatingActionButton = {
            if (uiState.activeVehicleId != null) {
                FloatingActionButton(
                    onClick = { onAddType(uiState.activeVehicleId!!) }
                ) {
                    Icon(
                        imageVector = Icons.Default.Add,
                        contentDescription = "Add maintenance type"
                    )
                }
            }
        }
    ) { innerPadding ->
        when {
            uiState.isLoading -> {
                Box(
                    modifier = Modifier
                        .fillMaxSize()
                        .padding(innerPadding),
                    contentAlignment = Alignment.Center
                ) {
                    CircularProgressIndicator()
                }
            }

            uiState.activeVehicleId == null -> {
                NoVehicleState(
                    modifier = Modifier
                        .fillMaxSize()
                        .padding(innerPadding)
                )
            }

            uiState.statuses.isEmpty() -> {
                EmptyMaintenanceState(
                    modifier = Modifier
                        .fillMaxSize()
                        .padding(innerPadding)
                )
            }

            else -> {
                LazyColumn(
                    modifier = Modifier
                        .fillMaxSize()
                        .padding(innerPadding),
                    contentPadding = PaddingValues(16.dp),
                    verticalArrangement = Arrangement.spacedBy(12.dp)
                ) {
                    items(
                        items = uiState.statuses,
                        key = { it.type.id }
                    ) { status ->
                        SwipeToDismissMaintenanceCard(
                            status = status,
                            onClick = { onViewHistory(status.type.id) },
                            onEdit = {
                                onEditType(uiState.activeVehicleId!!, status.type.id)
                            },
                            onDismiss = {
                                val deletedType = status.type
                                viewModel.deleteType(deletedType)
                                scope.launch {
                                    val result = snackbarHostState.showSnackbar(
                                        message = "${deletedType.name} deleted",
                                        actionLabel = "Undo",
                                        duration = SnackbarDuration.Short
                                    )
                                    if (result == SnackbarResult.ActionPerformed) {
                                        // Re-insert is not a perfect undo (new id), but functional
                                    }
                                }
                            }
                        )
                    }
                }
            }
        }
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun SwipeToDismissMaintenanceCard(
    status: MaintenanceStatus,
    onClick: () -> Unit,
    onEdit: () -> Unit,
    onDismiss: () -> Unit
) {
    val dismissState = rememberSwipeToDismissBoxState(
        confirmValueChange = { dismissValue ->
            if (dismissValue == SwipeToDismissBoxValue.EndToStart) {
                onDismiss()
                true
            } else {
                false
            }
        }
    )

    SwipeToDismissBox(
        state = dismissState,
        backgroundContent = {
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(vertical = 4.dp),
                contentAlignment = Alignment.CenterEnd
            ) {
                Icon(
                    imageVector = Icons.Default.Delete,
                    contentDescription = "Delete",
                    modifier = Modifier.padding(end = 24.dp),
                    tint = MaterialTheme.colorScheme.error
                )
            }
        },
        enableDismissFromStartToEnd = false
    ) {
        MaintenanceTypeCard(
            status = status,
            onClick = onClick,
            onEdit = onEdit
        )
    }
}

@Composable
private fun MaintenanceTypeCard(
    status: MaintenanceStatus,
    onClick: () -> Unit,
    onEdit: () -> Unit
) {
    val urgencyColor = when (status.urgency) {
        MaintenanceUrgency.OK -> Color(0xFF4CAF50)
        MaintenanceUrgency.DUE_SOON -> Color(0xFFFFC107)
        MaintenanceUrgency.OVERDUE -> Color(0xFFF44336)
    }

    val urgencyLabel = when (status.urgency) {
        MaintenanceUrgency.OK -> "OK"
        MaintenanceUrgency.DUE_SOON -> "Due Soon"
        MaintenanceUrgency.OVERDUE -> "Overdue"
    }

    val urgencyTextColor = when (status.urgency) {
        MaintenanceUrgency.OK -> Color.White
        MaintenanceUrgency.DUE_SOON -> Color.Black
        MaintenanceUrgency.OVERDUE -> Color.White
    }

    Card(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(onClick = onClick)
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp)
        ) {
            // Name and urgency badge
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Text(
                    text = status.type.name,
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.SemiBold,
                    modifier = Modifier.weight(1f)
                )
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    Badge(
                        containerColor = urgencyColor,
                        contentColor = urgencyTextColor
                    ) {
                        Text(
                            text = urgencyLabel,
                            modifier = Modifier.padding(horizontal = 4.dp)
                        )
                    }
                    IconButton(
                        onClick = onEdit,
                        modifier = Modifier.size(32.dp)
                    ) {
                        Icon(
                            imageVector = Icons.Default.Edit,
                            contentDescription = "Edit ${status.type.name}",
                            modifier = Modifier.size(18.dp)
                        )
                    }
                }
            }

            Spacer(modifier = Modifier.height(8.dp))

            // Miles until due or overdue
            Text(
                text = if (status.milesUntilDue >= 0) {
                    "${FormatUtil.formatMiles(status.milesUntilDue)} until due"
                } else {
                    "Overdue by ${FormatUtil.formatMiles(abs(status.milesUntilDue))}"
                },
                style = MaterialTheme.typography.bodyMedium,
                color = if (status.urgency == MaintenanceUrgency.OVERDUE) {
                    MaterialTheme.colorScheme.error
                } else {
                    MaterialTheme.colorScheme.onSurfaceVariant
                }
            )

            Spacer(modifier = Modifier.height(8.dp))

            // Progress bar
            LinearProgressIndicator(
                progress = { status.percentUsed.coerceIn(0.0, 1.0).toFloat() },
                modifier = Modifier
                    .fillMaxWidth()
                    .height(8.dp),
                color = urgencyColor,
                trackColor = urgencyColor.copy(alpha = 0.2f)
            )

            // Last serviced date
            if (status.lastServiceDate != null) {
                Spacer(modifier = Modifier.height(8.dp))
                Text(
                    text = "Last serviced: ${FormatUtil.formatDate(status.lastServiceDate)}",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
        }
    }
}

@Composable
private fun NoVehicleState(modifier: Modifier = Modifier) {
    Box(
        modifier = modifier,
        contentAlignment = Alignment.Center
    ) {
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.Center
        ) {
            Icon(
                imageVector = Icons.Default.Build,
                contentDescription = null,
                modifier = Modifier.size(64.dp),
                tint = MaterialTheme.colorScheme.onSurfaceVariant
            )
            Spacer(modifier = Modifier.height(16.dp))
            Text(
                text = "Select a vehicle first",
                style = MaterialTheme.typography.titleLarge,
                color = MaterialTheme.colorScheme.onSurface
            )
            Spacer(modifier = Modifier.height(8.dp))
            Text(
                text = "Set an active vehicle from the Vehicles tab to track maintenance",
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                textAlign = TextAlign.Center,
                modifier = Modifier.padding(horizontal = 32.dp)
            )
        }
    }
}

@Composable
private fun EmptyMaintenanceState(modifier: Modifier = Modifier) {
    Box(
        modifier = modifier,
        contentAlignment = Alignment.Center
    ) {
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.Center
        ) {
            Icon(
                imageVector = Icons.Default.Build,
                contentDescription = null,
                modifier = Modifier.size(64.dp),
                tint = MaterialTheme.colorScheme.onSurfaceVariant
            )
            Spacer(modifier = Modifier.height(16.dp))
            Text(
                text = "No maintenance items",
                style = MaterialTheme.typography.titleLarge,
                color = MaterialTheme.colorScheme.onSurface
            )
            Spacer(modifier = Modifier.height(8.dp))
            Text(
                text = "Set up your first maintenance item",
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                textAlign = TextAlign.Center
            )
        }
    }
}

File: vtrack-ios/VTrack/Views/Maintenance/MaintenanceTypesScreen.swift
import SwiftUI
import SwiftData

struct MaintenanceTypesScreen: View {
    @Environment(\.modelContext) private var modelContext
    @State private var viewModel: MaintenanceTypesViewModel?

    var onAddType: (String) -> Void
    var onEditType: (String, String) -> Void
    var onViewHistory: (String) -> Void

    var body: some View {
        Group {
            if let vm = viewModel {
                if vm.activeVehicle == nil {
                    EmptyStateView(icon: "car.fill", title: "No Vehicle", message: "Add a vehicle first to track maintenance")
                } else if vm.statuses.isEmpty {
                    EmptyStateView(icon: "wrench.and.screwdriver.fill", title: "No maintenance types", message: "Tap + to add your first maintenance schedule")
                } else {
                    List {
                        ForEach(vm.statuses, id: \.type.persistentModelID) { status in
                            MaintenanceTypeRow(status: status) {
                                let vehicleId = vm.activeVehicle!.persistentModelID.hashValue.description
                                let typeId = status.type.persistentModelID.hashValue.description
                                onEditType(vehicleId, typeId)
                            }
                            .contentShape(Rectangle())
                            .onTapGesture {
                                onViewHistory(status.type.persistentModelID.hashValue.description)
                            }
                        }
                        .onDelete { indexSet in
                            for index in indexSet {
                                vm.deleteType(vm.statuses[index].type)
                            }
                        }
                    }
                }
            } else {
                ProgressView()
            }
        }
        .toolbar {
            ToolbarItem(placement: .primaryAction) {
                Button {
                    if let vehicleId = viewModel?.activeVehicle?.persistentModelID.hashValue.description {
                        onAddType(vehicleId)
                    }
                } label: {
                    Image(systemName: "plus")
                }
                .disabled(viewModel?.activeVehicle == nil)
            }
        }
        .onAppear {
            if viewModel == nil {
                viewModel = MaintenanceTypesViewModel(modelContext: modelContext)
            }
            viewModel?.load()
        }
    }
}

struct MaintenanceTypeRow: View {
    let status: MaintenanceStatus
    let onEdit: () -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack {
                Text(status.type.name)
                    .font(.headline)
                urgencyBadge
                Spacer()
                Button(action: onEdit) {
                    Image(systemName: "pencil")
                        .foregroundStyle(.secondary)
                }
                .buttonStyle(.plain)
            }
            Text(statusText)
                .font(.subheadline)
                .foregroundStyle(.secondary)
            ProgressView(value: min(max(status.percentUsed, 0), 1))
                .tint(urgencyColor)
            if let date = status.lastServiceDate {
                Text("Last serviced: \(FormatUtil.formatDate(date))")
                    .font(.caption)
                    .foregroundStyle(.secondary)
            }
        }
        .padding(.vertical, 4)
    }

    private var statusText: String {
        switch status.urgency {
        case .overdue:
            return "Overdue by \(FormatUtil.formatMiles(abs(status.milesUntilDue)))"
        case .dueSoon:
            return "\(FormatUtil.formatMiles(status.milesUntilDue)) remaining"
        case .ok:
            return "\(FormatUtil.formatMiles(status.milesUntilDue)) remaining"
        }
    }

    @ViewBuilder
    private var urgencyBadge: some View {
        let (text, bg, fg): (String, Color, Color) = {
            switch status.urgency {
            case .overdue: return ("Overdue", .red, .white)
            case .dueSoon: return ("Due Soon", .orange, .black)
            case .ok: return ("OK", .green, .white)
            }
        }()
        Text(text)
            .font(.caption2)
            .fontWeight(.semibold)
            .padding(.horizontal, 6)
            .padding(.vertical, 2)
            .background(bg)
            .foregroundStyle(fg)
            .clipShape(Capsule())
    }

    private var urgencyColor: Color {
        switch status.urgency {
        case .overdue: return .red
        case .dueSoon: return .orange
        case .ok: return .green
        }
    }
}
